create @lavamoat/types

boneskull opened the pull request:
Closes #995

This creates a new package, `@lavamoat/types`, which provides common types for various packages. It solves a type-only dependency cycle between `lavamoat-core` and `lavamoat-tofu`, at minimum

* * *

**UPDATE May 28 2025**:  The _scope_ of this PR is:

- Resolve the cycle by creating `@lavamoat/types`, populated with types from `lavamoat-core`
- Consume `@lavamoat/types` where `lavamoat-core` was previously consumed (for the appropriate types)
- Use `@import` syntax in JS sources (though I do not claim to have addressed _all_ of these)
- Avoid using deprecated `LavaMoatOverrides` type
- Narrow types if appropriate (this is the only place where types would be meaningfully different, though I expect any such changes to be backwards-compatible)

_Changes to logic are explicitly **not** in scope._

boneskull committed:
feat(types): create @lavamoat/types

# Conflicts:
#	package-lock.json

boneskull committed:
fix(tofu,core,node,survey): move codeSampleFromAstNode to core

...and use @lavamoat/types

boneskull committed:
chore: update depcheck cfgs

# Conflicts:
#	packages/browserify/.depcheckrc.yml

legobeat reviewed (commented):
@boneskull needs conflict-resolution of:

```
package-lock.json
packages/tofu/package.json 
```

legobeat commented on packages/core/src/moduleRecord.js:
```diff
@@ -1,77 +1,38 @@
 // @ts-check
 
+/**
+ * Do not import this; use `import('@lavamoat/types').LavamoatModuleRecord`
```
Considering that this is importing the same (so should be no risk of falling out of sync): Is there any danger in importing this one instead?

After reading into it, it feels kind of like I wouldn't need to know any of this and while not ideal, importing from here should be OK?

(Once the types package has landed properly, I'm thinking we could also look at if there's some linting rule we could have, that could in general prevent direct imports directly from files in other workspace packages?)

legobeat replied:
Also: `@deprecated` might fit here?
https://jsdoc.app/tags-deprecated

boneskull replied:
It's not deprecated.  So the point is that we need to use `LavamoatModuleRecord` from the `types` package as an interface and implement it.  

In JSDoc--and I think this is a bug--`@implements` works _if and only if_ the literal type you're using is a `@typedef` _in the current module_.  That means:

```js
/** @typedef {import('./butts').Butts} Butts */

/**
 * @implements {Butts}
 */
class Pants {
}
``` 

works fine, but:

```js
/**
 * @implements {import('./butts').Butts}
 */
class Pants {
}
``` 

_does not_.


But if we use a `@typedef`, we re-export the type, because that's what typedefs do.  But we only want one source of truth for the `LavamoatModuleRecord` interface.  Because we have to use the `@typedef`, we've now created a "copy" of `LavamoatModuleRecord` named `LMR`.  Nobody should use it.

That said, if we can upgrade TS, we can probably avoid the `@typedef` altogether because of the new `@import` tag.

legobeat commented on packages/types/src/module-record.ts:
Could these be `.d.ts` files directly, rather than `.ts`?

boneskull replied:
No, because `.d.ts` files are generally not typechecked.

legobeat replied:
Only because we are inheriting from a base tsconfig which sets `compilerOption.skipLibCheck = true` (which I only realized now).

It will check them if set to `false` (which I really think we should in general): 
https://github.com/LavaMoat/LavaMoat/pull/1264/commits/82e966be469edc56ffbb9e71b97ebb7d958ce91d

(BTW: Setting the same configuration in `packages/types/tsconfig.json` did not take effect during `build:types`. This was surprising to me)

boneskull replied:
No, I don't recommend that.  It will cause things to break if one of our dependencies has bad types.

boneskull replied:
At any rate: this is a package full of types, and IMO TypeScript sources are more than appropriate here.

legobeat replied:
> It will cause things to break if one of our dependencies has bad types.

.. In which case it will be caught by type-checking before it becomes an actual problem, right?

As opposed to without, in which case it will still affect and break any downstream users who do enable full type-checking, without us noticing.

legobeat replied:
> At any rate: this is a package full of types, and IMO TypeScript sources are more than appropriate here.

`.d.ts` files are guaranteed to only include types sans implementation. Reduces the potential for future fun times because someone somehow puts an `enum` or `class` there.

boneskull replied:
> .. In which case it will be caught by type-checking before it becomes an actual problem, right?

> As opposed to without, in which case it will still affect and break any downstream users who do enable full type-checking, without us noticing.

1. It will break, but we won't be able to fix it, because we do not own those types.
2. Most people don't enable this level of type-checking. We trust that our dependencies ship good types. If they don't ship good types, well, that sucks, but we're not gonna break the build because of it.

> .d.ts files are guaranteed to only include types sans implementation. Reduces the potential for future fun times because someone somehow puts an enum or class there.

I really don't understand the concern here.  This is a package for types.

legobeat replied:
> I really don't understand the concern here. This is a package for types.

`.ts` files can include and export things that are not types (TS `enum` is just an example of something people commonly believe is a type, but is actually also runtime code, a confusion which commonly causes great consternation).

I think the types package should include type interfaces only - no resulting js code - and `.d.ts` files reflect that and makes it more naturally enforcable.

legobeat replied:
> It will break, but we won't be able to fix it, because we do not own those types.

We can, though? Do you have an example of an unfixable issue? If we pull in and rely on types beyond our control, I think that's a bigger issue in itself?

boneskull committed:
fix(webpack): add @lavamoat/types

boneskull committed:
fix(types): do not generate .js files

- add `exports` field

boneskull committed:
chore(ts): reorder root tsconfig.json references

boneskull committed:
fix(types): update package.json and lockfile

boneskull committed:
fix(core): bad type in mergePolicy

boneskull committed:
fix(core,node): updates post-merge of @lavamoat/node

- corrected a param type in `endowmentsToolkit()`
- corrected constructor param type in `LavamoatModuleRecord` class
- pull all the types out of `@lavamoat/types`

^ Conflicts:
^	packages/core/src/moduleRecord.js
^	packages/node/src/exec/run.js
^	packages/node/src/policy-converter.js
^	packages/node/src/policy-gen/generate.js
^	packages/node/src/policy-gen/policy-gen-compartment-class.js
^	packages/node/src/policy-gen/to-policy.js
^	packages/node/src/policy-util.js
^	packages/node/src/types.ts
^	packages/node/test/unit/policy/conversion.spec.js
^	packages/node/test/unit/policy/policy-macros.js
^	packages/node/test/unit/policy/types.ts

boneskull committed:
fix(webpack): remove now-unused @ts-expect-error; add project ref to @lavamoat/types

boneskull committed:
feat(types): provide Object.hasOwn definition

This copies the `Object.hasOwn` type out of core

boneskull committed:
fix(types): Object.hasOwn does not interfere with index signatures

The `Object.hasOwn` type was doing a thing where if you have object `foo` of type `Record<string, any>` and you asked if `Object.hasOwn(foo, 'bar')`, TS would infer that the _value_ of `foo[bar]` is `never` -- which means you cannot subsequently _assign a value_ to `foo[bar]` within a block.

This fixes that problem, but the type is still broken in some circumstances (see note I added to definition).

boneskull committed:
chore(core): remove has-own type def in lieu of @lavamoat/types

boneskull committed:
chore(core): convert all inline type imports to @import statements

boneskull committed:
fix(tofu): update type for new SetFieldType behavior

boneskull committed:
fix(types): update @babel/types

boneskull committed:
chore(types): add lint:deps script

boneskull committed:
chore(types): remove reference to non-existent @types/ses

boneskull committed:
chore(core): avoid deprecated LavaMoatPolicyOverrides type

boneskull committed:
fix(webpack): fix type import

boneskull committed:
chore: update lockfile

naugtur commented on packages/tofu/src/index.js:
```diff
@@ -25,5 +24,4 @@ module.exports = {
   traverse,
   inspectPrimordialAssignments,
   inspectSesCompat,
-  codeSampleFromAstNode,
```
@boneskull this was a breaking change
